whitepapers/0013_ccq.md

```diff
@@ -0,0 +1,350 @@
+
+```go
+u8       version
+u32      nonce
+u8       num_per_chain_queries
+[]byte   per_chain_queries
+```
+
+### Per-Chain Query
+
+Multiple queries for the same chain may be submitted in a single `Per-Chain Query`.
+
+```go
+u16      chain_id
+u8       type
+uint32   query_len
+[]byte   query_data
+```
+
+Currently the only supported query type is `eth_call`. This can be expanded to support other protocols.
+
+1. eth_call (query type 1)
+
+   ```go
+   u32      block_id_len
+   []byte   block_id
+   u8       num_batch_call_data
+   []byte   batch_call_data
+   ```
+
+   Calls are batched to allow specifying multiple calls against the same block. These will be done in the same batch RPC call and are easier for the requestor to verify on chain.
+
+   ```go
+   [20]byte   contract_address
+   u32        call_data_len
+   []byte     call_data
+   ```
+
+## Query Response
+
+- Off-Chain
+  ```go
+  u8         version
+  u16        sender_chain_id = 0
+  [65]byte   signature
+  u32        query_request_len
+  []byte     query_request
+  u8         num_per_chain_responses
+  []byte     per_chain_responses
+  ```
+- On-Chain [WIP] - depends on whether the request is done via VAA or not, this could be chain/emitter/sequence but that wouldn’t work with faster-than-finality
+  ```go
+  u16        sender_chain_id != 0
+  [32]byte   vaa_hash
+  []byte     query_request
+  u8         num_per_chain_responses
+  []byte     per_chain_responses
+  ```
+- Per-Chain Responses
+
+  ```go
+  u16     chain_id
+  u8      type
+  uint32  response_len
+  []byte  response
+  ```
+
+  1. eth_call (query type 1) Response Body
+
+     ```go
+     u64         block_number
+     [32]byte    block_hash
+     u64         block_time_us
+     u8          num_results
+     []byte      results
+     ```
+
+     ```go
+     u32         result_len
+     []byte      result
+     ```
+
+## REST Service
+
+### Request
+
+```ts
+export interface QueryRequest {
+  bytes: string; // As a hex string
+  signatures?: string; // As a hex string
+}
+```
+
+Note that the signature may be omitted if the user is provisioned to allow unsigned requests in the CCQ server configuration. In that case,
+the CCQ server will sign the request.
+
+#### Request Headers
+
+The requests must include the following header.
+
+- "X-API-Key" must match what is configured on the CCQ server.
+
+### Response
+
+- 200 - request reached consensus in < 1m
+
+  - Signatures is the ECDSA signature + a byte for the index of the guardian in the guardian set
+  - Signatures will appear in increasing set index order
+
+  ```ts
+  export interface QueryResponse {
+    bytes: string; // As a hex string
+    signatures?: string; // As a hex string
+  }
+  ```
+
+- 400 - bad request (malformed input or bytes)
+- 403 - [future] e.g. requestor doesn’t have an account set up
+- 500 - [future] failed to reach consensus (e.g. received 14 responses but 7 with one result and 7 with another)
+- 504 - did not reach consensus in < 1m
```

Comment:
3 issues I see / the motivation for this design:

- The server here _is_ acting as a gateway to the guardian network (but we already have too many things in the space called gateways). Since it is not responsible for servicing the request, and there is purposefully no response for the gossip network on bad / failed requests at this time (to reduce the negative effects on the gossip network), the server cannot return accurate information on a failure past "the service I'm proxying didn't return a response".
- We were explicitly asked to hold open the request until the response was received. This is basically a long-poll design. Switching to short-polling as you suggest would likely require more memory consumption on the server and more management in the front-end around these requests. Short-polling would also require advanced considerations for load balancing and rate limiting. In the current design, more CCQ servers can be stood up equally servicing requests, but if switched to your suggestion they would require sticky load balancing or a shared database backing requests/responses.
- I don't see `503` as appropriate. As described above, it does not refer to the unavailability of the service, but rather the fact that a response was not returned. The only part of the 504 use that I question is that, in _many_ (but not all) cases, a 504 cannot be resolved by the end user. Here, however, the user _might_ be able to change their request to get a response.

Reply:
Scratch that, this makes sense for the first take.

Reply:
Absolutely! The current testnet CCQ server is actually placed behind an application load balancer (as you expect) and I would anticipate anyone running this publicly facing would treat this as a (potentially) one-of-many backend service to be load-balanced and proxied :slightly_smiling_face: